// @aws-sdk/rds-signer/dist-cjs/index.js
'use strict';

var util = require('@aws-sdk/core/util');
var protocols = require('@smithy/core/protocols');
var signatureV4 = require('@smithy/signature-v4');
var runtimeConfig = require('./runtimeConfig');

const MINUTE_MS = 60_000;
class Signer {
    credentials;
    hostname;
    port;
    protocol = "https:";
    region;
    service = "rds-db";
    sha256;
    username;
    constructor(configuration) {
        const runtimeConfiguration = runtimeConfig.getRuntimeConfig(configuration);
        const creds = runtimeConfiguration.credentials;
        this.credentials = typeof creds === "function" ? this.createCredentialsWrapper(creds) : creds;
        this.hostname = runtimeConfiguration.hostname;
        this.port = runtimeConfiguration.port;
        this.region = runtimeConfiguration.region;
        this.sha256 = runtimeConfiguration.sha256;
        this.username = runtimeConfiguration.username;
    }
    async getAuthToken() {
        const signer = new signatureV4.SignatureV4({
            service: this.service,
            region: this.region,
            credentials: this.credentials,
            sha256: this.sha256,
        });
        const request = new protocols.HttpRequest({
            method: "GET",
            protocol: this.protocol,
            hostname: this.hostname,
            port: this.port,
            query: {
                Action: "connect",
                DBUser: this.username,
            },
            headers: {
                host: `${this.hostname}:${this.port}`,
            },
        });
        const presigned = await signer.presign(request, {
            expiresIn: 900,
        });
        return util.formatUrl(presigned).replace(`${this.protocol}//`, "");
    }
    createCredentialsWrapper(provider) {
        return async (identityProperties) => {
            const credentials = await provider(identityProperties);
            if (credentials.expiration && credentials.expiration.getTime() - Date.now() < 15 * MINUTE_MS) {
                try {
                    const refreshed = await provider({ ...identityProperties, forceRefresh: true });
                    if (!refreshed.expiration || refreshed.expiration.getTime() - Date.now() >= 15 * MINUTE_MS) {
                        return refreshed;
                    }
                }
                catch { }
            }
            return credentials;
        };
    }
}

exports.Signer = Signer;
